src/meshes/t8code_mesh.jl

```diff
@@ -0,0 +1,383 @@
+    #   if !T8code.uses_mpi()
+    #     error("t8code library does not support MPI")
+    #   end
+    #   is_parallel = Val(true)
+    # else
+    #   is_parallel = Val(false)
+    # end
+    is_parallel = False()
+
+    mesh = new{NDIMS, Float64, typeof(is_parallel)}(cmesh, scheme, forest, is_parallel)
+
+    # Destroy 't8code' structs when the mesh is garbage collected.
+    finalizer(function (mesh::T8codeMesh{NDIMS})
+      trixi_t8_unref_forest(mesh.forest)
+      finalize_t8code()
+    end, mesh)
+
+    return mesh
+  end
+end
+
+function T8codeMesh{NDIMS}(cmesh, scheme, forest, tree_node_coordinates, nodes, boundary_names,
+                          current_filename, unsaved_changes) where NDIMS
+
+  mesh = T8codeMesh{NDIMS}(cmesh, scheme, forest)
+
+  mesh.nodes = nodes
+  mesh.boundary_names = boundary_names
+  mesh.unsaved_changes = unsaved_changes
+  mesh.current_filename = current_filename
+  mesh.tree_node_coordinates = tree_node_coordinates
+
+  return mesh
+end
+
+const SerialT8codeMesh{NDIMS} = T8codeMesh{NDIMS, <:Real, <:False}
+@inline mpi_parallel(mesh::SerialT8codeMesh) = False()
+
+@inline Base.ndims(::T8codeMesh{NDIMS}) where NDIMS = NDIMS
+@inline Base.real(::T8codeMesh{NDIMS, RealT}) where {NDIMS, RealT} = RealT
+
+# TODO: What should be returned in case of parallel processes? Local vs global.
+@inline ntrees(mesh::T8codeMesh) = Int(t8_forest_get_num_global_trees(mesh.forest))
+# @inline ncells(mesh::T8codeMesh) = Int(t8_forest_get_global_num_elements(mesh.forest))
+@inline ncells(mesh::T8codeMesh) = Int(t8_forest_get_local_num_elements(mesh.forest))
+@inline ninterfaces(mesh::T8codeMesh) = mesh.ninterfaces
+@inline nmortars(mesh::T8codeMesh) = mesh.nmortars
+@inline nboundaries(mesh::T8codeMesh) = mesh.nboundaries
+
+function Base.show(io::IO, mesh::T8codeMesh)
+  print(io, "T8codeMesh{", ndims(mesh), ", ", real(mesh), "}")
+end
+
+function Base.show(io::IO, :: MIME"text/plain", mesh::T8codeMesh)
+  if get(io, :compact, false)
+    show(io, mesh)
+  else
+    setup = [
+             "#trees" => ntrees(mesh),
+             "current #cells" => ncells(mesh),
+             "polydeg" => length(mesh.nodes) - 1,
+            ]
+    summary_box(io, "T8codeMesh{" * string(ndims(mesh)) * ", " * string(real(mesh)) * "}", setup)
+  end
+end
+
+"""
+    T8codeMesh(trees_per_dimension; polydeg,
+              mapping=nothing, faces=nothing, coordinates_min=nothing, coordinates_max=nothing,
+              RealT=Float64, initial_refinement_level=0, periodicity=true, unsaved_changes=true)
+Create a structured curved 'T8codeMesh' of the specified size.
+There are three ways to map the mesh to the physical domain.
+1. Define a 'mapping' that maps the hypercube '[-1, 1]^n'.
+2. Specify a 'Tuple' 'faces' of functions that parametrize each face.
+3. Create a rectangular mesh by specifying 'coordinates_min' and 'coordinates_max'.
+Non-periodic boundaries will be called ':x_neg', ':x_pos', ':y_neg', ':y_pos', ':z_neg', ':z_pos'.
+# Arguments
```

Comment:
```suggestion
1. Define a `mapping` that maps the hypercube '[-1, 1]^n'.
2. Specify a 'Tuple' 'faces' of functions that parametrize each face.
3. Create a rectangular mesh by specifying 'coordinates_min' and 'coordinates_max'.

Non-periodic boundaries will be called ':x_neg', ':x_pos', ':y_neg', ':y_pos', ':z_neg', ':z_pos'.

# Arguments
```
Did you check the rendering of all docstrings? You need backticks `` ` `` to render something as code, not `'`.